src/handlers/motions/motionFinalized/helpers.ts

```diff
@@ -96,3 +116,84 @@ export const linkPendingDomainMetadataWithDomain = async (action: string, colony
     });
   }
 };
+
+export const linkPendingColonyMetadataWithColony = async (
+  pendingColonyMetadata: ColonyMetadata,
+  colonyAddress: string,
+): Promise<void> => {
+  const currentColonyMetadata = await query<ColonyMetadata>(
+    'getColonyMetadata',
+    {
+      id: colonyAddress,
+    },
+  );
+
+  if (!currentColonyMetadata) {
+    console.error(
+      `Could not find the current metadata for the colony: ${colonyAddress}. This is a bug and should be investigated.`,
+    );
+    return;
+  }
+
+  const {
+    haveTokensChanged,
+    hasAvatarChanged,
+    hasWhitelistChanged,
+    newDisplayName,
+    oldDisplayName,
+  } = pendingColonyMetadata.changelog?.[0] ?? {};
+
+  const updatedMetadata = {
+    ...currentColonyMetadata,
+  };
+
+  /*
+   * Here, we update metadata as granularly as possible so that we don't overwrite state changes that occured
+   * after this motion was created.
+   */
+
+  if (hasAvatarChanged) {
+    // If avatar has changed, update avatar and thumbnail
+    updatedMetadata.avatar = pendingColonyMetadata.avatar;
+    updatedMetadata.thumbnail = pendingColonyMetadata.thumbnail;
+  }
+
+  if (hasWhitelistChanged) {
+    // If whitelist has changed, update whitelistedAddresses and isWhitelistActivated
+    updatedMetadata.isWhitelistActivated =
+      pendingColonyMetadata.isWhitelistActivated;
+
+    updatedMetadata.whitelistedAddresses =
+      pendingColonyMetadata.whitelistedAddresses;
+  }
+
+  if (newDisplayName !== oldDisplayName) {
+    // If displayName has changed, update displayName
+    updatedMetadata.displayName = pendingColonyMetadata.displayName;
+  }
+
+  if (haveTokensChanged && pendingColonyMetadata.modifiedTokenAddresses) {
+    // If tokens have changed, update colony tokens
+    const colony = await getColonyFromDB(colonyAddress);
+
+    if (colony) {
+      const existingTokenAddresses = getExistingTokenAddresses(colony);
+
+      await updateColonyTokens(
```

Comment:
This logic is taken from the `editColony` saga in the CDapp. It's the same, except we are running it only once a successful motion has been finalized.